contracts/StorageFees.cdc

```diff
@@ -0,0 +1,244 @@
+            pre {
+                self.verify(): "StorageReservation not on owners account or not on the correct path"
+            }
+            if from.balance == 0.0 {
+                destroy from // it is empty so we can destroy it
+                return
+            }
+            let oldStorageReservation = self.reservedTokens.balance
+            
+            self.reservedTokens.deposit(from: <- from)
+
+            emit StorageReservationChanged(
+                address: self.ownerAddress, 
+                oldStorageReservation: oldStorageReservation, 
+                oldStorageCapacity: StorageFees.flowToStorageCapacity(oldStorageReservation), 
+                newStorageReservation: self.reservedTokens.balance, 
+                newStorageCapacity: StorageFees.flowToStorageCapacity(self.reservedTokens.balance))
+        }
+
+        // The `withdraw` function allows the owner of this resource to withdraw flow tokens from it if the owner decides that he/she/it doesn't need as much storage capacity any more.
+        pub fun withdraw(amount: UFix64): @FlowToken.Vault {
+            pre {
+                self.verify(): "StorageReservation not on owners account or not on the correct path"
+                StorageFees.refundingEnabled: "Refunding is currently disabled"
+                self.reservedTokens.balance - amount >= StorageFees.minimumStorageReservation:  "Cannot withdraw below the minimum storage reservation"
+            }
+            if amount == UFix64(0.0) {
+                return <- (FlowToken.createEmptyVault() as! @FlowToken.Vault)
+            }
+            let oldStorageReservation = self.reservedTokens.balance
+            
+            let vault <- (self.reservedTokens.withdraw(amount: amount) as! @FlowToken.Vault)
+
+            emit StorageReservationChanged(
+                address: self.ownerAddress, 
+                oldStorageReservation: oldStorageReservation, 
+                oldStorageCapacity: StorageFees.flowToStorageCapacity(oldStorageReservation), 
+                newStorageReservation: self.reservedTokens.balance, 
+                newStorageCapacity: StorageFees.flowToStorageCapacity(self.reservedTokens.balance))
+
+            return <- vault
+        }
+
+        // Verify itself, that it is on the expected account on the expected path and of the expected type.
+        access(self) fun verify(): Bool {
+            let receiver = StorageFees.getStorageReservationReceiver(self.ownerAddress)
+            return receiver.verifyStorageReservation(ownerAddress: self.ownerAddress, storageReservationId: self.storageReservationId)
+        }
+
+        // This is called from the method above. The StorageReservation on this location should be the one that is expected.
+        access(contract) fun verifyStorageReservation(ownerAddress: Address, storageReservationId: UInt64): Bool {
+            if self.ownerAddress != ownerAddress || self.storageReservationId != self.storageReservationId  {
```

Comment:
Should there be a `self.owner.address!` somewhere here? It looks like this will never fail because the owner address that is passed to it from `verify` is always the same as `self.ownerAddress`

Reply:
I will add more description to this function to try to make it cleared how it functions.